File: 1. Two Sum.java
// solution 1: double loop
// Time: n^2
class Solution {
    public int[] twoSum(int[] nums, int target) {
        // 两层循环
        for (int i = 0; i < nums.length - 1; i++) {
            for (int j = i+1; j < nums.length; j++) {
                if (nums[i] + nums[j] == target) {
                    // https://stackoverflow.com/questions/1200621/how-do-i-declare-and-initialize-an-array-in-java
                    return new int[] {i, j};
                }
            }
        }
        // https://www.w3schools.com/java/ref_keyword_throw.asp
        throw new IllegalArgumentException("No two sum solution");
    }
}


// solution 2: hashing，两次遍历，第一次创建hashmap，第二次对每个元素进行查询
// Time: n，但是需要额外空间存储hashmap，复杂度为n
class Solution {
    public int[] twoSum(int[] nums, int target) {
        // 创建hashmap并放入数据
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            // put(key, value)
            map.put(nums[i], i);
        }
        // 遍历查询
        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];
            // HashMap.containsKey(): true/false
            // HashMap.get(): index
            if (map.containsKey(complement) && map.get(complement) != i) {
                return new int[] { i, map.get(complement) };
            }
        }
    throw new IllegalArgumentException("No two sum solution");
    }
}

File: 1. Two Sum.py
# solution 1: double loop
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        for i in range(0, len(nums)-1):
            for j in range(i+1,len(nums)):
                if nums[i]+nums[j] == target:
                    return [i,j]
        return None


# solution 2: hashmap
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # use dictionary as hashmap
        # https://www.w3schools.com/python/python_dictionaries.asp
        Map = {}
        for i in range(0, len(nums)):
            Map[nums[i]] = i
        # search for complement
        for i in range(0, len(nums)):
            complement = target - nums[i]
            # 记得检查两个数的下标是否相同
            if complement in Map and Map[complement] != i:
                return [i, Map[complement]]
        return None


# solution 3: hashmap one pass
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        Map = {}
        for i in range(len(nums)):
            complement = target - nums[i]
            if complement in Map and Map[complement] != i:
                return [i, Map[complement]]
            else:
                Map[nums[i]] = i
        return None


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        hashmap = dict()
        for index in range(len(nums)):
            left = target - nums[index]
            if left in hashmap:
                return [index, hashmap[left]]
            else:
                hashmap[nums[index]] = index
        
